metadata/src/main/java/org/apache/kafka/migration/KRaftMigrationDriver.java

```diff
@@ -0,0 +1,414 @@
+            log.error("Had an exception in " + this.getClass().getSimpleName(), e);
+        }
+    }
+
+    class MigrateMetadataEvent implements EventQueue.Event {
+        @Override
+        public void run() throws Exception {
+            if (migrationState != MigrationState.ZK_MIGRATION) {
+                log.warn("Skipping ZK migration, already done");
+                return;
+            }
+
+            Set<Integer> brokersWithAssignments = new HashSet<>();
+            log.info("Begin migration from ZK");
+            consumer.beginMigration();
+            try {
+                // TODO use a KIP-868 metadata transaction here
+                List<CompletableFuture<?>> futures = new ArrayList<>();
+                client.readAllMetadata(batch -> futures.add(consumer.acceptBatch(batch)), brokersWithAssignments::add);
+                CompletableFuture.allOf(futures.toArray(new CompletableFuture[]{})).get();
+
+                Set<Integer> brokersWithRegistrations = new HashSet<>(zkBrokerRegistrations.keySet());
+                brokersWithAssignments.removeAll(brokersWithRegistrations);
+                if (!brokersWithAssignments.isEmpty()) {
+                    //throw new IllegalStateException("Cannot migrate data with offline brokers: " + brokersWithAssignments);
+                    log.error("Offline ZK brokers detected: {}", brokersWithAssignments);
+                }
+
+                // Update the migration state
+                OffsetAndEpoch offsetAndEpoch = consumer.completeMigration();
+                apply("Migrating ZK to KRaft", __ -> migrationState(offsetAndEpoch.offset, offsetAndEpoch.epoch));
+            } catch (Throwable t) {
+                log.error("Migration failed", t);
+                consumer.abortMigration();
+            } finally {
+                // TODO Just skip to dual write for now
+                apply("Persist recovery state", client::setMigrationRecoveryState);
+                transitionTo(MigrationState.DUAL_WRITE);
+            }
+        }
+
+        @Override
+        public void handleException(Throwable e) {
+            log.error("Had an exception in " + this.getClass().getSimpleName(), e);
+        }
+    }
+
+    class BrokersChangeEvent implements EventQueue.Event {
+        @Override
+        public void run() throws Exception {
+            Set<Integer> updatedBrokerIds = client.readBrokerIds();
+            Set<Integer> added = new HashSet<>(updatedBrokerIds);
+            added.removeAll(zkBrokerRegistrations.keySet());
+
+            Set<Integer> removed = new HashSet<>(zkBrokerRegistrations.keySet());
+            removed.removeAll(updatedBrokerIds);
+
+            log.debug("ZK Brokers added: " + added + ", removed: " + removed);
+            added.forEach(brokerId -> {
+                Optional<ZkBrokerRegistration> broker = client.readBrokerRegistration(brokerId);
+                if (broker.isPresent()) {
+                    client.addZkBroker(brokerId);
```

Comment:
I think the client addition of the broker should be done as the final step of ZkBrokerRegistration reconciliation in KRaft controller. Once all the validations are done, I think we should first add BrokerRegistrationChangeRecord for the addition or removal of Zk broker. Also, validation itself is probably an async step since we decided to use hearbeats to see if brokers are ready.

Reply:
Technically I think this method doesn't need to directly talk to ClusterLinkControlManager if we do the above IMO.